src/rust_option.rs

```diff
@@ -0,0 +1,527 @@
+    }
+
+    /// SAFETY: Pointer must not be aliased
+    pub unsafe fn as_mut_option_mut<'a>(&mut self) -> &mut Option<&'a mut T> {
+        let _: () = assert_option_safe::<&mut T>();
+        unsafe { &mut *(self as *mut RustOption<*mut T> as *mut Option<&'a mut T>) }
+    }
+}
+
+impl RustOption<*const core::ffi::c_void> {
+    pub fn from_option_ref_improper<'a, T>(other: Option<&'a T>) -> Self {
+        let _: () = assert_option_safe::<&T>();
+        unsafe {
+            core::mem::transmute::<Option<&'a T>, RustOption<*const core::ffi::c_void>>(other)
+        }
+    }
+
+    pub fn from_option_ref_improper_pinned<'a, T>(other: Option<Pin<&'a T>>) -> Self {
+        let _: () = assert_option_safe::<Pin<&T>>();
+        unsafe {
+            core::mem::transmute::<Option<Pin<&'a T>>, RustOption<*const core::ffi::c_void>>(other)
+        }
+    }
+
+    /// SAFETY: Pointer must have been constructed as a Box
+    #[cfg(feature = "alloc")]
+    pub unsafe fn into_option_box_improper<T>(self) -> Option<Box<T>> {
+        let _: () = assert_option_safe::<Box<T>>();
+        unsafe {
+            core::mem::transmute::<RustOption<*const core::ffi::c_void>, Option<Box<T>>>(self)
+        }
+    }
+
+    /// SAFETY: Pointer must have been constructed as a Box
+    #[cfg(feature = "alloc")]
+    pub unsafe fn as_mut_option_box_improper<T>(&mut self) -> &mut Option<Box<T>> {
+        let _: () = assert_option_safe::<Box<T>>();
+        unsafe { &mut *(self as *mut RustOption<*const core::ffi::c_void> as *mut Option<Box<T>>) }
+    }
+
+    pub fn into_option_ref_improper<'a, T>(self) -> Option<&'a T> {
+        let _: () = assert_option_safe::<&T>();
+        unsafe { core::mem::transmute::<RustOption<*const core::ffi::c_void>, Option<&'a T>>(self) }
+    }
+
+    pub fn as_mut_option_ref_improper<'a, T>(&mut self) -> &mut Option<&'a T> {
+        let _: () = assert_option_safe::<&T>();
+        unsafe { &mut *(self as *mut RustOption<*const core::ffi::c_void> as *mut Option<&'a T>) }
+    }
+}
+
+impl RustOption<*mut core::ffi::c_void> {
+    #[cfg(feature = "alloc")]
+    pub fn from_option_box_improper<T>(other: Option<Box<T>>) -> Self {
+        let _: () = assert_option_safe::<Box<T>>();
+        unsafe { core::mem::transmute::<Option<Box<T>>, RustOption<*mut core::ffi::c_void>>(other) }
+    }
+
+    pub fn from_option_mut_improper<'a, T>(other: Option<&'a mut T>) -> Self {
+        let _: () = assert_option_safe::<&mut T>();
+        unsafe {
+            core::mem::transmute::<Option<&'a mut T>, RustOption<*mut core::ffi::c_void>>(other)
+        }
+    }
+
+    pub fn from_option_mut_improper_pinned<'a, T>(other: Option<Pin<&'a mut T>>) -> Self {
+        let _: () = assert_option_safe::<Pin<&mut T>>();
+        unsafe {
+            core::mem::transmute::<Option<Pin<&'a mut T>>, RustOption<*mut core::ffi::c_void>>(
+                other,
+            )
+        }
+    }
+
+    /// SAFETY: Pointer must have been constructed as a Box
```

Comment:
I'd expand on these safety comments slightly. `self` must have been constructed as a `Option<Box<T>>`. Later on, the "Pointer must not be aliased and must have been constructed as an Option<&mut T>"